File: python/src/agent_framework/core/run_manager.py
from __future__ import annotations

from dataclasses import replace
from datetime import datetime, timezone
from uuid import uuid4

from agent_framework.core.contracts import RunManager, RunRecord


class DefaultRunManager(RunManager):
    def __init__(self) -> None:
        self._runs: dict[str, RunRecord] = {}

    def create(self, session_id: str) -> RunRecord:
        record = RunRecord(
            run_id=str(uuid4()),
            session_id=session_id,
            status="pending",
            started_at=datetime.now(timezone.utc).isoformat(),
        )
        self._runs[record.run_id] = record
        return record

    def update(self, run_id: str, patch: dict) -> None:
        existing = self._runs.get(run_id)
        if existing is None:
            return
        self._runs[run_id] = replace(existing, **patch)

    def get(self, run_id: str) -> RunRecord | None:
        return self._runs.get(run_id)

File: typescript/packages/core/src/run-manager.ts
import { randomUUID } from 'node:crypto';
import type { RunManager, RunRecord } from './types/contracts.js';
import type { RunStatus } from './types/domain.js';

export class DefaultRunManager implements RunManager {
  private runs = new Map<string, RunRecord>();
  private waiters = new Map<string, Array<(record: RunRecord) => void>>();

  create(sessionId: string): RunRecord {
    const record: RunRecord = {
      runId: randomUUID(),
      sessionId,
      status: 'pending',
      startedAt: new Date().toISOString(),
    };
    this.runs.set(record.runId, record);
    return record;
  }

  update(runId: string, patch: Partial<RunRecord>): void {
    const existing = this.runs.get(runId);
    if (!existing) {
      return;
    }
    const updated = { ...existing, ...patch };
    this.runs.set(runId, updated);
    if (isTerminal(updated.status)) {
      for (const resolve of this.waiters.get(runId) ?? []) {
        resolve(updated);
      }
      this.waiters.delete(runId);
    }
  }

  get(runId: string): RunRecord | undefined {
    return this.runs.get(runId);
  }

  wait(runId: string, timeoutMs = 600_000): Promise<RunRecord> {
    const existing = this.runs.get(runId);
    if (existing && isTerminal(existing.status)) {
      return Promise.resolve(existing);
    }
    return new Promise<RunRecord>((resolve, reject) => {
      const timeout = setTimeout(() => reject(new Error(`Run ${runId} timed out`)), timeoutMs);
      const list = this.waiters.get(runId) ?? [];
      list.push((record) => {
        clearTimeout(timeout);
        resolve(record);
      });
      this.waiters.set(runId, list);
    });
  }
}

function isTerminal(status: RunStatus): boolean {
  return status === 'completed' || status === 'failed' || status === 'cancelled';
}
